### PR TITLE
add timestamp offset && server stream read service

### PR DESCRIPTION
**Pull request by YangKian**

### CODE REVIEW

File: hstream.proto [hunk 1]
```diff
@@ -128,10 +129,16 @@ enum SpecialOffset {
...
+message TimestampOffset {
+  int64 timestampInMs = 1;
+  bool strictAccuracy = 2;
+}
+
 message ShardOffset {
   oneof offset {
     SpecialOffset specialOffset = 1;
     RecordId recordOffset = 2;
+    TimestampOffset timestampOffset = 3;
```
Comment (4eUeP) [inline hstream.proto:141]: Does this have a default offset in the server?

Comment (YangKian) [reply hstream.proto:141]: No, offset is specified in the request sent by the client, server just gets offset from rpc request

File: hstream.proto [hunk 2]
```diff
@@ -346,6 +354,16 @@ message ReadShardResponse { repeated ReceivedRecord receivedRecords = 1; }
 message ListShardReadersRequest {}
 message ListShardReadersResponse { repeated string readerId = 1; }
 
+message ReadShardStreamRequest {
+  string readerId = 1;
+  uint64 shardId = 2;
+  ShardOffset shardOffset = 3;
+  uint32 timeout = 4;
+  uint32 maxRecords = 5;
```
Comment (YangKian) [inline hstream.proto:362]: `maxRecords` is a parameter required by the underlying storage read interface, indicating that a read call waits for a maximum of `maxRecords` data before returning. Note it does not indicate that the current shardReader will only read a maximum of `maxRecords` data.

Comment (4eUeP) [reply hstream.proto:362]: How about renaming `maxRecords` to `maxBatched`?

Comment (YangKian) [reply hstream.proto:362]: maybe `maxBatchedPerRead` or `maxBatchedWaited`?

**Review by 4eUeP — APPROVED** [review]

**Pull request merged**